### DISCUSSION TITLE
filter() function  - weird issue - detailed description for Deltic

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5095] [Q&A Core] [answered]
Hi,
site one is posting data to another site and received a response.  e107 is used as the backend for the WP frontend.  

WP uses this for the return value:
`https://developer.wordpress.org/reference/functions/get_term_link/` 
As you see, it returns either a string or the object.

The response is created this way:  (was, there was wrong error code)
 
```
$response = [
                        'code' => "1",
                        'message' =>  $taxonomy .  "doslo k chybe",
                        'importId' => $term_id,
                        'url' => get_term_link($term_id),
                    ];
```
 
Now, e107 part. 

```
                $response = curl_exec($curl);
                curl_close($curl);
                $result = json_decode($response);

                if ($result->code == 1)
                {
                    $url = $result->url;
                    $url = e107::getParser()->filter($url, "str");
                }
```

Error:
Fatal error: Uncaught Error: Object of class stdClass could not be converted to string in ***\ehandlers\e_parse_class.php on line 5176
Error: Object of class stdClass could not be converted to string in ***\ehandlers\e_parse_class.php on line 5176

```
Call Stack
#	Time	Memory	Function	Location
1	0.0016	376048	{main}( )	...\admin_fandom.php:0
2	0.1576	1610448	e_admin_dispatcher->__construct( $auto_observe = ???, $request = ???, $response = ??? )	...\admin_fandom.php:271
3	0.1581	1615360	e_admin_dispatcher->runObservers( $run_header = ??? )	...\admin_ui.php:1109
4	0.1585	1628104	e_admin_controller->dispatchObserver( $action = ??? )	...\admin_ui.php:1348
5	0.1748	1706656	e_admin_ui->ListBatchTrigger( $batch_trigger = 'options__sync_with_wp' )	...\admin_ui.php:2252
6	0.1748	1707032	e_admin_controller_ui->_handleListBatch( $batch_trigger = 'options__sync_with_wp' )	...\admin_ui.php:5431
7	0.1756	1708048	fanfiction_categories_ui->handleListOptionsBatch( $selected = [0 => '152', 1 => '142', 2 => '141', 3 => '140', 4 => '139', 5 => '138', 6 => '23'], $type = 'sync_with_wp' )	...\admin_ui.php:3870
8	0.1756	1708104	fanfiction_categories_ui->sync_with_wp( $ids = '152,142,141,140,139,138,23' )	...\admin_fandom.php:191
9	0.9238	1731824	e_parse->filter( $text = class stdClass { public $errors = class stdClass { public $invalid_term = [...] }; public $error_data = [] }, $type = 'str', $validate = ??? )	...\admin_fandom.php:164
10	0.9238	1733392	e_parse->{closure:C:\wamp64\www\***\ehandlers\e_parse_class.php:5174-5179}( $element = class stdClass { public $errors = class stdClass { public $invalid_term = [...] }; public $error_data = [] } )	...\e_parse_class.php:5186


```

**Comment by Deltik** [comment]
Hey @Jimmi08,

Your issue isn't within the e107 core.  The `$url` is a stdClass object that looks like this:

```
class stdClass { public $errors = class stdClass { public $invalid_term = [...] }; public $error_data = [] }
```

But `e_parse::filter()` takes a string or an array.

Your external web request is responding with something unexpected, so you should look into that.

**Reply by Jimmi08** [reply]
The response is correct.  it is either a string or an object.

https://developer.wordpress.org/reference/functions/get_term_link/
string|[WP_Error](https://developer.wordpress.org/reference/classes/wp_error/) URL of the taxonomy term archive on success, [WP_Error](https://developer.wordpress.org/reference/classes/wp_error/) if term does not exist.

I know that I can fix this and I did it, but it surprised me that this crashed site.

**Reply by Deltik** [reply, marked as answer]
Your glue code is responsible for ensuring the right argument type(s).  Otherwise, the behavior is undefined, as you could be passing bad data into the e107 core.

Here's a diagram showing the types that your glue code is juggling:

![image](https://github.com/e107inc/e107/assets/1364268/19a1a891-8344-476a-984d-91cb5179eedb)

It just so happens that your JSON object sometimes gives you the `string` type to pass to `e_parse::filter()`, but there could also be a `stdClass` that is not acceptable to `e_parse::filter()`.  You should perform error handling in your glue code to ensure that the `e_parse::filter()` is used correctly.
